### PR TITLE
GNUIndent: Add more configs

### PR DESCRIPTION
**Pull request by Asnelchristian**
Closes https://github.com/coala-analyzer/coala-bears/issues/373

### CODE REVIEW

File: bears/c_languages/GNUIndentBear.py [hunk 1]
```diff
@@ -23,18 +23,125 @@ class GNUIndentBear:
...
+            changes to
+            ```
+            if (...) {
+                ....
+            } else {
+            ```
+
+        :param cuddle_do_while:
+            Cuddles while of ``do {} while``; and preceding ``}``.
+        :param space_special_semicolon:
```
Comment (Adrianzatreanu) [inline bears/c_languages/GNUIndentBear.py]: param name is so not helpful

Comment (Asnelchristian) [inline bears/c_languages/GNUIndentBear.py]: ``cuddle_do_while`` right ? Thinking of another name for it, any suggestion? :)

Comment (Adrianzatreanu) [inline bears/c_languages/GNUIndentBear.py]: no, i meant ``space_special_semicolon``

**Comment by Asnelchristian** [comment]
I also want to know which of the options you think are useless, that will be helpfull ;)

File: bears/c_languages/GNUIndentBear.py (hunk 1 above)
Comment (Adrianzatreanu) [inline bears/c_languages/GNUIndentBear.py]: maybe ``space_before_semicolon``? not sure

Comment (Asnelchristian) [inline bears/c_languages/GNUIndentBear.py]: ah okay, Agreed `space_before_semicolon ` is better :+1:

File: bears/c_languages/GNUIndentBear.py [hunk 2]
```diff
@@ -23,18 +23,125 @@ class GNUIndentBear:
...
+        :param ignore_newlines:
+            Does not break long lines at the position of newlines in the input
+            file.
+        :param declaration_indent:
+            Puts variables in column (declaration_indent).
+        :param brace_indent:
+            Indents braces (brace_indent) spaces.
+        :param continue_at_parentheses:
+            Lines up continued lines at parentheses.
+        :param continuation_indentation:
```
Comment (Adrianzatreanu) [inline bears/c_languages/GNUIndentBear.py]: i fail to understand what this option does from its description

File: bears/c_languages/GNUIndentBear.py [hunk 3]
```diff
@@ -23,18 +23,125 @@ class GNUIndentBear:
...
+            Forces newline after comma in declaration.
+        :param ignore_newlines:
+            Does not break long lines at the position of newlines in the input
+            file.
+        :param declaration_indent:
+            Puts variables in column (declaration_indent).
+        :param brace_indent:
+            Indents braces (brace_indent) spaces.
+        :param continue_at_parentheses:
+            Lines up continued lines at parentheses.
```
Comment (Adrianzatreanu) [inline bears/c_languages/GNUIndentBear.py]: also pretty bad description. descriptions should be helpful, not causing more ambiguity

File: bears/c_languages/GNUIndentBear.py [hunk 4]
```diff
@@ -23,18 +23,125 @@ class GNUIndentBear:
...
+            ``while`` statements.
+        :param blank_lines_after_commas:
+            Forces newline after comma in declaration.
+        :param ignore_newlines:
+            Does not break long lines at the position of newlines in the input
+            file.
+        :param declaration_indent:
+            Puts variables in column (declaration_indent).
+        :param brace_indent:
+            Indents braces (brace_indent) spaces.
```
Comment (Adrianzatreanu) [inline bears/c_languages/GNUIndentBear.py]: so what does this do exactly? ``(    text    )`` instead of ``(text)``? cuz thats what it says and i dont know why youd want that

File: bears/c_languages/GNUIndentBear.py [hunk 5]
```diff
@@ -23,18 +23,125 @@ class GNUIndentBear:
...
+        :param space_special_semicolon:
+            Forces a blank before the semicolon ``;`` on one-line ``for`` and
+            ``while`` statements.
+        :param blank_lines_after_commas:
+            Forces newline after comma in declaration.
+        :param ignore_newlines:
+            Does not break long lines at the position of newlines in the input
+            file.
+        :param declaration_indent:
+            Puts variables in column (declaration_indent).
```
Comment (Adrianzatreanu) [inline bears/c_languages/GNUIndentBear.py]: so this places every variable on a new line? `` puts them in column `` ?

File: bears/c_languages/GNUIndentBear.py [hunk 6]
```diff
@@ -23,18 +23,125 @@ class GNUIndentBear:
...
+            else {
+            ```
+            changes to
+            ```
+            if (...) {
+                ....
+            } else {
+            ```
+
+        :param cuddle_do_while:
```
Comment (Adrianzatreanu) [inline bears/c_languages/GNUIndentBear.py]: i'd change ``cuddle`` as a term. i understood what it does, but honestly it doesnt fit here ^^

**Comment by Adrianzatreanu** [comment]
all in all, PR looks great, but honestly, there are many ambiguous descriptions that i dont understand or that are unhelpful instead of helping understand what they do.

File: bears/c_languages/GNUIndentBear.py (hunk 5 above)
Comment (Asnelchristian) [inline bears/c_languages/GNUIndentBear.py]: not really, `declaration_indent` is a number if its value is 16 for example:
```
  int               a_num;
  char           *a_char_ptr;
```
the variable names all start at column 16

Comment (Asnelchristian) [inline bears/c_languages/GNUIndentBear.py]: i I undrestand what you are saying .
I will improve on the description of the parameters

Comment (Adrianzatreanu) [inline bears/c_languages/GNUIndentBear.py]: i get it now, improve it, thanks :D

File: bears/c_languages/GNUIndentBear.py [hunk 7]
```diff
@@ -23,18 +23,155 @@ class GNUIndentBear:
...
-                                   understands. They will be simply passed
-                                   through.
+        :param max_line_length:
+            Maximum number of characters for a line.
+        :param use_spaces:
+            True if spaces are to be used, else tabs.
+        :param blank_lines_after_declarations:
+            Forces blank lines after the declarations.
+
+            Example: If ``True``,
```
Comment (abhsag24) [inline bears/c_languages/GNUIndentBear.py]: so it'd be better if the example starts on a new line and is indented so that the formatting works and we get a better output on the rst

Comment (Asnelchristian) [inline bears/c_languages/GNUIndentBear.py]: Agreed :+1:

File: bears/c_languages/GNUIndentBear.py [hunk 8]
```diff
@@ -1,5 +1,5 @@
 import platform
-import shlex
+from shlex import split
```
Comment (Makman2) [inline bears/c_languages/GNUIndentBear.py]: please keep the old one, we need to make some discussion out of this anyway, but as split is a builtin for string, it's better understandable together with the module name
```
"a b".split()
# vs
shlex.split("a b")
```

**Comment by hypothesist** [comment]
@Asnelchristian Use `Closes` instead of `Fixes` in the commit message? Since this is a feature request more than a bugfix.

**Comment by Asnelchristian** [comment]
i will do that

File: bears/c_languages/GNUIndentBear.py [hunk 9]
```diff
@@ -23,18 +23,160 @@ class GNUIndentBear:
...
+                         blank_lines_after_declarations: bool=False,
+                         blank_lines_after_procedures: bool=False,
+                         blank_lines_after_commas: bool=False,
+                         braces_on_if_line: bool=False,
+                         braces_on_func_def_line: bool=False,
+                         cuddle_else: bool=False,
+                         while_and_brace_on_same_line: bool=False,
+                         case_indentation: int=0,
+                         space_before_semicolon_after_empty_loop: bool=True,
+                         delete_optional_blank_lines: bool=False,
```
Comment (sils1297) [inline bears/c_languages/GNUIndentBear.py]: I'd make this a true

Comment (Asnelchristian) [inline bears/c_languages/GNUIndentBear.py]: right :+1:

2016-06-16 20:44 GMT+01:00 Lasse Schuirmann <notifications@github.com>:

File: bears/c_languages/GNUIndentBear.py [hunk 10]
```diff
@@ -23,18 +23,160 @@ class GNUIndentBear:
...
+                           if declaration_indent != 0 else ())
+        indent_options += (("--case-indentation"+str(case_indentation),)
+                           if case_indentation != 0 else ())
+        indent_options += (("--space-special-semicolon",)
+                           if space_before_semicolon_after_empty_loop
+                           else ("-nss",))
+        indent_options += ("--brace-indent"+str(brace_indent),)
+        indent_options += ("--gnu-style",) if gnu_style else ()
+        indent_options += ("--k-and-r-style",) if k_and_r_style else ()
+        indent_options += ("--linux-style",) if linux_style else ()
```
Comment (sils1297) [inline bears/c_languages/GNUIndentBear.py]: hm, those style arguments are a combination of the others, I wonder if the defaults we have above override those potentially and thus do not result in the wanted style?

Comment (Asnelchristian) [inline bears/c_languages/GNUIndentBear.py]: if a option is provided explicilty, i will override the one provided by the
style

2016-06-16 20:47 GMT+01:00 Lasse Schuirmann <notifications@github.com>:

File: bears/c_languages/GNUIndentBear.py [hunk 11]
```diff
@@ -23,18 +23,166 @@ class GNUIndentBear:
...
-                                   through.
+        :param max_line_length:
+            Maximum number of characters for a line.
+        :param use_spaces:
+            True if spaces are to be used, else tabs.
+        :param blank_lines_after_declarations:
+            Forces blank lines after the declarations.
+
+            Example:
+                If ``True``,
```
Comment (abhsag24) [inline bears/c_languages/GNUIndentBear.py]: why not `IF`under triple backticks as well? any reason?

File: bears/c_languages/GNUIndentBear.py [hunk 12]
```diff
@@ -23,18 +23,166 @@ class GNUIndentBear:
...
+                return ...;
+                ```
+
+        :param blank_lines_after_procedures:
+            Force blank lines after procedure bodies.
+        :param blank_lines_after_commas:
+            Forces newline after comma in declaration.
+
+            Example:
+                If ``True``,
```
Comment (abhsag24) [inline bears/c_languages/GNUIndentBear.py]: same here?

File: bears/c_languages/GNUIndentBear.py [hunk 13]
```diff
@@ -23,18 +23,166 @@ class GNUIndentBear:
...
+                           ("-br",)) if braces_on_if_line else ("-bl",)
+        indent_style_option = ()
+        indent_style_option += ("--gnu-style",) if gnu_style else ()
+        indent_style_option += (("--k-and-r-style",)
+                                if k_and_r_style and indent_style_option is ()
+                                else ())
+        indent_style_option += (("--linux-style",)
+                                if linux_style and indent_style_option is ()
+                                else ())
+        return (indent_style_option if indent_style_option is not ()
```
Comment (abhsag24) [inline bears/c_languages/GNUIndentBear.py]: so we cant have `indent_style_option` and `indent_options` together? just asking.

File: bears/c_languages/GNUIndentBear.py [hunk 14]
```diff
@@ -23,18 +23,166 @@ class GNUIndentBear:
...
+        :param while_and_brace_on_same_line:
+            Cuddles while of ``do {} while``; and preceding ``}``.
+        :param case_indentation:
+            Specifies the number of spaces by which ``case`` in the ``switch``
+            are indented.
+        :param space_before_semicolon_after_empty_loop:
+            Forces a blank before the semicolon ``;`` on one-line ``for`` and
+            ``while`` statements.
+        :param delete_optional_blank_lines:
+             Deletes optional blank lines.
```
Comment (abhsag24) [inline bears/c_languages/GNUIndentBear.py]: whats an optional blank line?

File: bears/c_languages/GNUIndentBear.py (hunk 13 above)
Comment (Asnelchristian) [inline bears/c_languages/GNUIndentBear.py]: the problem is that  indent_options overrides the options provided by the
style_options, but you are right i did not solve the problem we were facing.
We don't  want the default options to override those of
`indent_style_options`

File: bears/c_languages/GNUIndentBear.py [hunk 15]
```diff
@@ -23,18 +23,166 @@ class GNUIndentBear:
...
+                changes to
+                ```
+                int a,
+                  b;
+                ```
+        :param braces_on_if_line:
+            Puts the brace ``{`` on same line with if.
+
+            Example:
+                If ``True``,
```
Comment (abhsag24) [inline bears/c_languages/GNUIndentBear.py]: same

File: bears/c_languages/GNUIndentBear.py (hunk 13 above)
Comment (Asnelchristian) [inline bears/c_languages/GNUIndentBear.py]: a blank line that is not needed, if you want for example want a blank line
after declarations of variables  then the line is needed

2016-06-17 12:56 GMT+01:00 Christian Ngoulla Sob <ngoullasob@gmail.com>:

> the problem is that  indent_options overrides the options provided by the
> style_options, but you are right i did not solve the problem we were facing.
> We don't  want the default options to override those of
> `indent_style_options`
>

Comment (abhsag24) [inline bears/c_languages/GNUIndentBear.py]: then please add that in the description?

File: bears/c_languages/GNUIndentBear.py [hunk 16]
```diff
@@ -23,18 +23,166 @@ class GNUIndentBear:
...
+        :param max_line_length:
+            Maximum number of characters for a line.
+        :param use_spaces:
+            True if spaces are to be used, else tabs.
+        :param blank_lines_after_declarations:
+            Forces blank lines after the declarations.
+
+            Example:
+                ```
+                If ``True``,
```
Comment (abhsag24) [inline bears/c_languages/GNUIndentBear.py]: i'm not sure the doubl backticks work from inside a triple backtick block? can you check that?

File: bears/c_languages/GNUIndentBear.py (hunk 13 above)
Comment (Asnelchristian) [inline bears/c_languages/GNUIndentBear.py]: I added a line before the return statement is that enough :)

2016-06-17 13:06 GMT+01:00 Abhay Raizada <notifications@github.com>:

File: bears/c_languages/GNUIndentBear.py [hunk 17]
```diff
@@ -23,18 +23,167 @@ class GNUIndentBear:
...
+                           ("-br",)) if braces_on_if_line else ("-bl",)
+        indent_style_option = ()
+        indent_style_option += ("--gnu-style",) if gnu_style else ()
+        indent_style_option += (("--k-and-r-style",)
+                                if k_and_r_style and indent_style_option is ()
+                                else ())
+        indent_style_option += (("--linux-style",)
+                                if linux_style and indent_style_option is ()
+                                else ())
+        # If a style is chosen the other configs aren't passed to `indent`
```
Comment (abhsag24) [inline bears/c_languages/GNUIndentBear.py]: nice :)

File: tests/c_languages/GNUIndentBearTest.py [hunk 18]
```diff
@@ -44,8 +100,43 @@
     settings={"use_spaces": "nope", "max_line_length": "80"})
 
 
-GNUIndentBearGNUIndentWidthTest = verify_local_bear(
+GNUIndentBearWidthTest = verify_local_bear(
```
Comment (abhsag24) [inline tests/c_languages/GNUIndentBearTest.py]: shouldnt this go in another commit?

File: bears/c_languages/GNUIndentBear.py (hunk 14 above)
Comment (abhsag24) [inline bears/c_languages/GNUIndentBear.py]: still need to expand docs here @Asnelchristian

File: tests/c_languages/GNUIndentBearTest.py (hunk 18 above)
Comment (Asnelchristian) [inline tests/c_languages/GNUIndentBearTest.py]: okay doing another commit  for that

2016-06-17 13:13 GMT+01:00 Abhay Raizada <notifications@github.com>:

File: bears/c_languages/GNUIndentBear.py [hunk 19]
```diff
@@ -23,18 +23,167 @@ class GNUIndentBear:
...
+        :param max_line_length:
+            Maximum number of characters for a line.
+        :param use_spaces:
+            True if spaces are to be used, else tabs.
+        :param blank_lines_after_declarations:
+            Forces blank lines after the declarations.
+
+            Example:
+                ```
+                If True,
```
Comment (abhsag24) [inline bears/c_languages/GNUIndentBear.py]: ok, so i got the true meaning of this line now, it states if the is setting is True and the rest of the code is after that, can you avoid this confusion? maybe you should put the IF Tue before the example or maybe reword its so that it is simpler, sorry for the confusion.

Comment (Asnelchristian) [inline bears/c_languages/GNUIndentBear.py]: i modified it a bit what do you think? :)

2016-06-17 13:23 GMT+01:00 Abhay Raizada <notifications@github.com>:

**Comment by abhsag24** [comment]
can you change the commit message to `Rename Test` ?

**Comment by Asnelchristian** [comment]
sure :)

2016-06-17 14:41 GMT+01:00 Abhay Raizada <notifications@github.com>:

File: bears/c_languages/GNUIndentBear.py [hunk 20]
```diff
@@ -23,18 +23,169 @@ class GNUIndentBear:
     def create_arguments(filename, file, config_file,
                          max_line_length: int=80,
                          use_spaces: bool=True,
+                         blank_lines_after_declarations: bool=False,
```
Comment (abhsag24) [inline bears/c_languages/GNUIndentBear.py]: do these follow any order?

Comment (Asnelchristian) [inline bears/c_languages/GNUIndentBear.py]: not really

2016-06-17 14:51 GMT+01:00 Abhay Raizada <notifications@github.com>:

**Comment by abhsag24** [comment]
unack 3f55f9b lowercase `t` please in `Test`

**Comment by Asnelchristian** [comment]
:+1:


2016-06-17 14:54 GMT+01:00 Abhay Raizada <notifications@github.com>:

File: bears/c_languages/GNUIndentBear.py (hunk 20 above)
Comment (abhsag24) [inline bears/c_languages/GNUIndentBear.py]: maybe it should, i guess its okay for now, maybe we should open an issue to order all params in lint bears alphabetically.

Comment (Asnelchristian) [inline bears/c_languages/GNUIndentBear.py]: okay :)

2016-06-17 14:56 GMT+01:00 Abhay Raizada <notifications@github.com>:

**Comment by abhsag24** [comment]
ack fa58f30

**Comment by abhsag24** [comment]
ack 4e61726

**Comment by Asnelchristian** [comment]
now all is left is to merge right? :)

2016-06-17 15:29 GMT+01:00 Abhay Raizada <notifications@github.com>:

**Comment by sils1297** [comment]
a rebase has to be done

**Comment by Asnelchristian** [comment]
reack https://github.com/coala-analyzer/coala-bears/pull/519/commits/e6a3408f69998e4a382840e4f92d2cc8d7e13fe2 https://github.com/coala-analyzer/coala-bears/pull/519/commits/714813f204aee788940cfdcc67a19e03f4148f91

**Comment by Asnelchristian** [comment]
reack https://github.com/coala-analyzer/coala-bears/pull/519/commits/e6a3408f69998e4a382840e4f92d2cc8d7e13fe2 https://github.com/coala-analyzer/coala-bears/pull/519/commits/714813f204aee788940cfdcc67a19e03f4148f91

**Comment by Adrianzatreanu** [comment]
@rultor merge

**Comment by Adrianzatreanu** [comment]
@rultor merge
